jung-visualization/src/main/java/edu/uci/ics/jung/visualization/spatial/SpatialGrid.java

```diff
@@ -287,49 +306,79 @@ public void recalculate(Collection<N> nodes) {
+    if (isActive()) {
+      if (!this.getLayoutArea().contains(location.x, location.y)) {
+        log.trace(location + " outside of spatial " + this.getLayoutArea());
+        this.setBounds(this.getUnion(this.getLayoutArea(), location.x, location.y));
+        recalculate(layoutModel.getGraph().nodes());
+      }
 
-    int rightBox = this.getBoxNumberFromLocation(nlayoutModel.apply(node));
-    // node should end up in box 'rightBox'
-    // check to see if it is already there
-    if (map.get(rightBox).contains(node)) {
-      // nothing to do here, just return
-      return;
-    }
-    // remove node from the first (and only) wrong box it is found in
-    Integer wrongBox = null;
-    synchronized (map) {
-      for (Integer box : map.keySet()) {
-        if (map.get(box).contains(node)) {
-          // remove it and stop, because node can be in only one box
-          wrongBox = box;
-          break;
+      int rightBox = this.getBoxNumberFromLocation(layoutModel.apply(node));
+      // node should end up in box 'rightBox'
+      // check to see if it is already there
+      if (map.get(rightBox).contains(node)) {
+        // nothing to do here, just return
+        return;
+      }
+      // remove node from the first (and only) wrong box it is found in
+      Integer wrongBox = null;
+      synchronized (map) {
+        for (Integer box : map.keySet()) {
+          if (map.get(box).contains(node)) {
+            // remove it and stop, because node can be in only one box
+            wrongBox = box;
+            break;
+          }
         }
       }
+      if (wrongBox != null) {
+        map.remove(wrongBox, node);
+      }
+      map.put(rightBox, node);
     }
-    if (wrongBox != null) {
-      map.remove(wrongBox, node);
-    }
-    map.put(rightBox, node);
   }
 
   @Override
   public N getClosestElement(Point2D p) {
-    return null;
+    return getClosestElement(p.getX(), p.getY());
   }
 
   @Override
   public N getClosestElement(double x, double y) {
+    if (!isActive()) {
+      return fallback.getNode(layoutModel, x, y);
+    }
+    Collection<TreeNode> leafs = getContainingLeafs(x, y);
+    if (leafs.size() != 0) {
+      TreeNode leaf = leafs.iterator().next();
+
+      Rectangle2D area = leaf.getBounds();
+      double radius = area.getWidth();
+      N closest = null;
+      while (closest == null) {
+
+        double diameter = radius * 2;
+
+        Ellipse2D searchArea = new Ellipse2D.Double(x - radius, y - radius, diameter, diameter);
+
+        Collection<N> nodes = getVisibleElements(searchArea);
+        closest = getClosest(nodes, x, y, radius);
+
+        // if I have already considered all of the nodes in the graph
+        // (in the spatialquadtree) there is no reason to enlarge the
+        // area and try again
+        if (nodes.size() >= layoutModel.getGraph().nodes().size()) {
+          break;
+        }
+        // double the search area size and try again
```

Comment:
documentation nit: you're quadrupling the search area size, not doubling it (because the size goes up as the square of the radius).  maybe just say "double the search radius" :)